demos/gplates_global/gplates_global.py

```diff
@@ -0,0 +1,263 @@
+import scipy.special
+import math
+rmin, rmax, ref_level, nlayers = 1.208, 2.208, 4, 8
+
+mesh2d = CubedSphereMesh(rmin, refinement_level=ref_level, degree=2)
+mesh = ExtrudedMesh(mesh2d, layers=nlayers, extrusion_type='radial')
+bottom_id, top_id = "bottom", "top"
+domain_volume = assemble(1*dx(domain=mesh))  # Required for a diagnostic calculation.
+
+V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
+W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+Q = FunctionSpace(mesh, "CG", 2)  # Temperature function space (scalar)
+Z = MixedFunctionSpace([V, W])  # Mixed function space.
+
+z = Function(Z)  # A field over the mixed function space Z.
+u, p = split(z)  # Returns symbolic UFL expression for u and p
+z.subfunctions[0].rename("Velocity")
+z.subfunctions[1].rename("Pressure")
+
+Ra = Constant(7e3)  # Rayleigh number
+approximation = BoussinesqApproximation(Ra)
+
+time = 0.0  # Initial time
+delta_t = Constant(1e-6)  # Initial time-step
+timesteps = 20  # Maximum number of timesteps
+t_adapt = TimestepAdaptor(delta_t, u, V, maximum_timestep=0.1, increase_tolerance=1.5)
+
+T = Function(Q, name="Temperature")
+T_avg = Function(Q, name='Layer_Averaged_Temp')
+T_dev = Function(Q, name='Temperature_Deviation')
+
+X = SpatialCoordinate(mesh)
+r = sqrt(X[0]**2 + X[1]**2 + X[2]**2)
+theta = atan2(X[1], X[0])  # Theta (longitude - different symbol to Zhong)
+phi = atan2(sqrt(X[0]**2+X[1]**2), X[2])  # Phi (co-latitude - different symbol to Zhong)
+
+conductive_term = rmin*(rmax - r) / (r*(rmax - rmin))
+l, m, eps_c, eps_s = 3, 2, 0.01, 0.01
+Plm = Function(Q, name="P_lm")
+cos_phi = Function(Q, name="cos_phi").interpolate(cos(phi))
+Plm.dat.data[:] = scipy.special.lpmv(m, l, cos_phi.dat.data_ro)  # Evaluate P_lm node-wise using scipy lpmv
+Plm.assign(Plm*math.sqrt(((2*l+1)*math.factorial(l-m))/(2*math.pi*math.factorial(l+m))))
+if m == 0:
+    Plm.assign(Plm/math.sqrt(2))
+T.interpolate(conductive_term +
+              (eps_c*cos(m*theta) + eps_s*sin(m*theta)) * Plm * sin(pi*(r - rmin)/(rmax-rmin)))
+
+averager = LayerAveraging(mesh, cartesian=False, quad_degree=6)
+averager.extrapolate_layer_average(T_avg, averager.get_layer_average(T))
+# -
+
+# ## Loading a 1-D Axisymmetric Profile for Viscosity
+# ----------
+# Earth's physical properties are primarily characterised by spherically symmetric (depth-dependent) features influenced predominantly by hydrostatic pressure variations.
+# Here, we load a 1-D viscosity profile, as utilised by Ghelichkhan et al. (2021) in *Geophysical Journal International* to model the Earth's evolution during the Cenozoic era.
+# We first set up our viscosity function space. The 1-D profile data is located in the file `./mu2_radial.rad`, and we will use interpolation functionalities provided by G-ADOPT to populate the viscosity field.
+
+mu = Function(Q, name="Viscosity")
+interpolate_1d_profile(function=mu, one_d_filename="mu2_radial.rad", cartesian=False)
```

Comment:
I tried running this, and get the following error:
```
Traceback (most recent call last):
  File "/home/firedrake/g-adopt/demos/gplates_global/gplates_global.py", line 80, in <module>
    interpolate_1d_profile(function=mu, one_d_filename="mu2_radial.rad", cartesian=False)
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/utility.py", line 514, in interpolate_1d_profile
    rad = Function(function.function_space()).interpolate(upward_coord)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/function.py", line 413, in interpolate
    return assemble(interp, tensor=self, ad_block_tag=ad_block_tag)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/assembly.py", line 30, in wrapper
    output = assemble(form, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 133, in assemble
    return get_assembler(expr, *args, **kwargs).assemble(tensor=tensor)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 381, in assemble
    result = BaseFormAssembler.base_form_postorder_traversal(self._form, visitor, visited)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 603, in base_form_postorder_traversal
    visited[e] = visitor(e, *(visited[arg] for arg in operands))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 377, in visitor
    return self.base_form_assembly_visitor(e, t, *operands)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 537, in base_form_assembly_visitor
    interpolator = firedrake.Interpolator(expression, expr.function_space(), **interp_data)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/pyadjoint/pyadjoint/tape.py", line 110, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/interpolation.py", line 819, in __init__
    self.callable, arguments = make_interpolator(expr, V, subset, access, bcs=bcs)
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/interpolation.py", line 998, in make_interpolator
    loops.extend(_interpolator(V, tensor, expr, subset, arguments, access, bcs=bcs))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<decorator-gen-31>", line 2, in _interpolator
  File "/home/firedrake/firedrake/src/firedrake/firedrake/utils.py", line 89, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/interpolation.py", line 1074, in _interpolator
    kernel = compile_expression(cell_set.comm, expr, to_element, V.ufl_element(),
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/PyOP2/pyop2/caching.py", line 293, in wrapper
    v = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/interpolation.py", line 1208, in compile_expression
    return compile_expression_dual_evaluation(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/tsfc/tsfc/driver.py", line 268, in compile_expression_dual_evaluation
    evaluation, basis_indices = to_element.dual_evaluation(fn)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/FInAT/finat/finiteelementbase.py", line 261, in dual_evaluation
    expr = sum_factorise(sum_indices, factors)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/tsfc/gem/optimise.py", line 358, in sum_factorise
    raise NotImplementedError("Too many indices for sum factorisation!")
NotImplementedError: Too many indices for sum factorisation!
```

Reply:
I guess `mu2_radial.rad` needs to be updated a) for the new radii; and b) for only 8 extruded layers?

Reply:
hmmm, I don't think it has to do with `mu2_radial`. It's a `interpolate_1d_profile` issue that goes back to changing to  `upward_component`. Somehow this works fine for `cartesian=True` (even for spherical case below). But does not work for `cartesian=False` :

```python3
from gadopt import *
from gadopt.utility import vertical_component, upward_normal

rmin, rmax, ref_level, nlayers = 1.208, 2.208, 4, 8
mesh2d = CubedSphereMesh(rmin, refinement_level=ref_level, degree=2)
mesh = ExtrudedMesh(mesh2d, layers=nlayers, extrusion_type='radial')

V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
Q = FunctionSpace(mesh, "CG", 2)  # Temperature function space (scalar)

upward_component = vertical_component(SpatialCoordinate(mesh), cartesian=False)

q = Function(Q, name="r")
q.interpolate(upward_component)
```

Reply:
This is now resolved via the recent PR.